### PR TITLE
Container Instance: Adding VNET support for container

### PR DESCRIPTION
**Pull request by samkreter**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py [hunk 1]
```diff
@@ -38,19 +39,36 @@ def _format_cpu_memory(container_group):
...
+    if ip_address:
+        return ip_address['type']
+
+
 def transform_container_group(result):
     """Transform a container group to table output. """
     return OrderedDict([('Name', result['name']),
                         ('ResourceGroup', result['resourceGroup']),
-                        ('ProvisioningState', result.get('provisioningState')),
+                        ('Status', _format_status(result)),
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py]: This would be a breaking change for anyone relying on the output format. Why this change?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py [hunk 2]
```diff
@@ -76,10 +76,16 @@ def load_arguments(self, _):
         c.argument('environment_variables', nargs='+', options_list=['--environment-variables', '-e'], type=_environment_variables_type, help='A list of environment variable for the container. Space-separated values in \'key=value\' format.')
         c.argument('secure_environment_variables', nargs='+', type=_secure_environment_variables_type, help='A list of secure environment variable for the container. Space-separated values in \'key=value\' format.')
         c.argument('secrets', secrets_type)
-        c.argument('network_profile', network_profile_type)
         c.argument('secrets_mount_path', validator=validate_volume_mount_path, help="The path within the container where the secrets volume should be mounted. Must not contain colon ':'.")
         c.argument('file', options_list=['--file', '-f'], help="The path to the input file.")
 
+    with self.argument_context('container create', arg_group='Network Arguments') as c:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: Should just say `arg_group='Network'`.  Otherwise this will say "Network Arguments Arguments" :)

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_validators.py [hunk 3]
```diff
@@ -37,6 +37,13 @@ def validate_gitrepo_directory(ns):
         raise CLIError("The git repo directory cannot contain '..'")
 
 
+def validate_subnet(ns):
+    if ns.vnet_name and not ns.subnet:
+        raise CLIError("'--subnet' is required when using  a virtual network.")
+    if ns.subnet and not ns.vnet_name:
+        raise CLIError("'--vnet-name' is required when using a virtual network.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_validators.py]: A more idiomatic error would look like:
`raise CLIError('usage error: --vnet-name NAME --subnet NAME | --subnet ID')`

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 4]
```diff
@@ -166,18 +174,22 @@ def create_container(cmd,
...
-
     # Concatenate secure and standard environment variables
     if environment_variables and secure_environment_variables:
         environment_variables = environment_variables + secure_environment_variables
     else:
         environment_variables = environment_variables or secure_environment_variables
 
+    # Set up VNET, subnet and network profile if needed
+    if subnet and vnet_name and not network_profile:
+        network_profile = _get_vnet_network_profile(cmd.cli_ctx, location, resource_group_name, vnet_name, vnet_address_prefix, subnet, subnet_address_prefix)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: If subnet ID is provided, vnet name should not be. I get the feeling this code still should be triggered in that instance?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 5]
```diff
@@ -202,6 +214,85 @@ def create_container(cmd,
...
+    aci_delegation = Delegation(
+        name="Microsoft.ContainerInstance.containerGroups",
+        service_name="Microsoft.ContainerInstance/containerGroups"
+    )
+
+    ncf = cf_network(cli_ctx)
+
+    subnet_name = subnet
+    if is_valid_resource_id(subnet):
+        subnet_name = parse_resource_id(subnet)['resource_name']
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: If subnet is a valid ID, you should probably also parse vnet name. In all other cases in the CLI, providing `--subnet ID --vnet-name NAME` raises an error.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 6]
```diff
@@ -202,6 +214,85 @@ def create_container(cmd,
...
+                                                                 location=location,
+                                                                 address_space=AddressSpace(address_prefixes=[vnet_address_prefix])))
+        subnet = Subnet(
+            name=subnet_name,
+            location=location,
+            address_prefix=subnet_address_prefix,
+            delegations=[aci_delegation])
+
+        subnet_poller = ncf.subnets.create_or_update(resource_group_name, vnet_name, subnet_name, subnet)
+        subnet = subnet_poller.result()
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: You can combine these.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py (hunk 1 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py]: It gives a more accurate view into the container to see if it is running or stopped. This should only effect the table output correct? which shouldn't break any apis

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py (hunk 2 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: Just caught that. Also is it best practice to put preview in this something like "Preview: Netowrk Arguments"?

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py (hunk 4 above)
Comment (samkreter) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Your correct. Making the change now.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py (hunk 2 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_params.py]: No, don't put preview.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py (hunk 1 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py]: okay that should be good.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_validators.py [hunk 7]
```diff
@@ -5,6 +5,8 @@
 
 from base64 import b64encode
 from knack.util import CLIError
+from azure.cli.core.commands.client_factory import get_subscription_id
+from msrestazure.tools import is_valid_resource_id, resource_id
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_validators.py]: This is your bad import. You must move this import inside the validator.

**Comment by troydai** [comment]
The CI is still failing in this PR.

**Comment by samkreter** [comment]
@troydai  Is  there any direct way to find the bad imports. I can't find anything pointing to the problem in the test

**Comment by tjprescott** [comment]
Sam, there's still a top-level import of msrestazure.tools in custom.py.

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py [hunk 8]
```diff
@@ -32,11 +32,13 @@
 from azure.mgmt.containerinstance.models import (AzureFileVolume, Container, ContainerGroup, ContainerGroupNetworkProtocol,
                                                  ContainerPort, ImageRegistryCredential, IpAddress, Port, ResourceRequests,
                                                  ResourceRequirements, Volume, VolumeMount, ContainerExecRequestTerminalSize,
-                                                 ContainerGroupIpAddressType, GitRepoVolume, LogAnalytics, ContainerGroupDiagnostics,
-                                                 ContainerGroupNetworkProfile)
+                                                 GitRepoVolume, LogAnalytics, ContainerGroupDiagnostics, ContainerGroupNetworkProfile,
+                                                 ContainerGroupIpAddressType)
+from azure.mgmt.network.models import (Subnet, VirtualNetwork, AddressSpace, Delegation, NetworkProfile,
+                                       ContainerNetworkInterfaceConfiguration, IPConfigurationProfile)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/custom.py]: Network is a multi-API module, so you can't do the imports like this.  In the relevant method, you need to do:

```Python
Subnet, VirtualNetwork = cmd.get_models('Subnet', 'VirtualNetwork')
```

This might be why you are getting the performance issue. Also, you can run the performance check locally with `azdev verify module-load-perf`.

**Comment by troydai** [comment]
The load performance of container regressed significantly

**Comment by troydai** [comment]
```
FAILED MODULES
Module                    Average    Threshold        Stdev                    Values
ALL                           365          300            2     [368.0, 363.0, 365.0]
container                     192           10            1     [194.0, 191.0, 192.0]
```

**Comment by samkreter** [comment]
@troydai is this caused only by the extra networking model imports?

**Comment by troydai** [comment]
From what I can see, it is likely the reason. One solution is to move the import statement closer to there it is actually used.

**Comment by samkreter** [comment]
All of the network models are only used in the one function to set up vnet. I can move the imports there and see if that helps the tests to pass.

**Comment by troydai** [comment]
Please also see @tjprescott 's last comments

File: src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 
 from __future__ import print_function
 from collections import OrderedDict
+from azure.mgmt.containerinstance.models import ContainerGroupIpAddressType
```
Comment (tjprescott) [inline src/command_modules/azure-cli-container/azure/cli/command_modules/container/_format.py]: This is the only other candidate I see for what could be causing the issue.
